core/src/main/java/io/grpc/internal/RetryPolicy.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.CallOptions;
+import io.grpc.MethodDescriptor;
+
+public abstract class RetryPolicy {
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+
+  RetryPolicy(BackoffPolicy.Provider backoffPolicyProvider) {
+    this.backoffPolicyProvider = backoffPolicyProvider;
+  }
+
+  RetryPolicy() {
+    this(new ExponentialBackoffPolicy.Provider());
+  }
+
+  public long getNextBackoffMillis() {
+    return backoffPolicyProvider.get().nextBackoffMillis();
+  }
+
+  public abstract boolean isRetryable(MethodDescriptor method, CallOptions callOptions);
```

Comment:
I think you also need the resulting `Status` of the call. For example, at Square with our current proprietary RPC framework, we have a policy that servers only send 503 (unavailable) and 429 (too many requests) if/when they have done zero work on behalf of the request. That way, clients are free to retry those status codes, even if the method in question is not idempotent.